experimental/gcp-log/internal/storage/storage.go

```diff
@@ -405,12 +430,34 @@ func (c *Client) StoreTile(ctx context.Context, level, index uint64, tile *api.T
 	tPath := filepath.Join(layout.TilePath("", level, index, tileSize%256))
 	obj := bkt.Object(tPath)
 
-	w := obj.NewWriter(ctx)
+	// Tiles, partial or full, should only be written once.
+	w := obj.If(gcs.Conditions{DoesNotExist: true}).NewWriter(ctx)
 	if c.otherCacheControl != "" {
 		w.ObjectAttrs.CacheControl = c.otherCacheControl
 	}
 	if _, err := w.Write(t); err != nil {
 		return fmt.Errorf("failed to write tile object %q to bucket %q: %w", tPath, c.bucket, err)
 	}
-	return w.Close()
+
+	if err := w.Close(); err != nil {
+		switch ee := err.(type) {
+		case *googleapi.Error:
+			// If we run into a precondition failure error, check that the object
+			// which exists contains the same content that we want to write.
+			if ee.Code == http.StatusPreconditionFailed {
+				if equal, err := c.assertContent(ctx, tPath, t); err != nil {
+					return fmt.Errorf("failed to read content of %q: %w", tPath, err)
+				} else if !equal {
+					return fmt.Errorf("assertion that tile content for %q has not changed failed", tPath)
+				} else if equal {
+					klog.V(2).Infof("StoreTile: Tile already exists for level %d index %x ts: %x", level, index, tileSize)
```

Comment:
super nit: worth adding "Identical" in the logging message somewhere for future us debugging some unrelated issue? (make it clearer this is not an error message, just unusual path informational message.)

Reply:
Done